sw/host/opentitanlib/src/proxy/handler.rs

```diff
@@ -0,0 +1,238 @@
+                        let mut resps: Vec<I2cTransferResponse> = reqs
+                            .iter()
+                            .map(|transfer| match transfer {
+                                I2cTransferRequest::Read { len } => I2cTransferResponse::Read {
+                                    data: vec![0; *len as usize],
+                                },
+                                I2cTransferRequest::Write { data: _ } => I2cTransferResponse::Write,
+                            })
+                            .collect();
+                        // Now carefully craft a proper parameter to the
+                        // `i2c::Bus::run_transactions()` method.  It will have reference
+                        // into elements of both the request vector and mutable reference into
+                        // the response vector.
+                        let mut transaction: Vec<i2c::Transfer> = reqs
+                            .iter()
+                            .zip(resps.iter_mut())
+                            .map(|pair| match pair {
+                                (
+                                    I2cTransferRequest::Read { len: _ },
+                                    I2cTransferResponse::Read { data },
+                                ) => i2c::Transfer::Read(data),
+                                (
+                                    I2cTransferRequest::Write { data },
+                                    I2cTransferResponse::Write,
+                                ) => i2c::Transfer::Write(&data),
+                                _ => {
+                                    // This can only happen if the logic in this method is
+                                    // flawed.  (Never due to network input.)
+                                    panic!("Mismatch");
+                                }
+                            })
+                            .collect();
+                        instance.run_transaction(*address, &mut transaction)?;
+                        Ok(Response::I2c(I2cResponse::RunTransaction {
+                            transaction: resps,
+                        }))
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl<'a> CommandHandler<Message> for TransportCommandHandler<'a> {
+    /// This method will perform whatever action on the underlying `Transport` that is requested
+    /// by the given `Message`, and return a response to be sent to the client.  Any `Err`
+    /// return from this method will be treated as an irrecoverable protocol error, causing an
+    /// error message in the server log, and the connection to be terminated.
+    fn execute_cmd(&self, msg: &Message) -> Result<Message> {
+        match msg {
+            // Package either `Ok()` or `Err()` into a `Message`, to be sent via network.
+            Message::Req(req) => Ok(Message::Res(self.do_execute_cmd(&req))),
```

Comment:
I think an `if let` would look nicer here:

```rust
if let Message::Req(req) = msg {
  return ...
}
bail!(...);
```

Up to you.

Reply:
You are right, this is slightly nicer. Done.